backend/penndata/views.py

```diff
@@ -268,6 +270,125 @@ def get_queryset(self):
+            after = snapshots.filter(date__gte=hour_date).order_by("date").first()
+
+            before_val = getattr(before, field, None)
+            after_val = getattr(after, field, None)
+
+            if before_val is None or after_val is None:
+                usage[hour] = self.safe_add(before_val, after_val)
+                # set to None if latest entry was more than an hour ago (to avoid extrapolation)
+                if (
+                    after_val is None
+                    and before
+                    and hour_date - datetime.timedelta(hours=1) > before.date
+                ):
+                    usage[hour] = None
+            else:
+                # linear interpolation
+                usage[hour] = (
+                    before_val
+                    + (after_val - before_val)
+                    * (hour_date - before.date).total_seconds()
+                    / (after.date - before.date).total_seconds()
+                )
+
+        if all(x == 0 for x in usage):  # location probably closed - don't count in aggregate
+            return [None] * 24
+        return usage
+
+    def get_usage(self, room, date, per_hour, group_by, field):
+        unit = 1 if group_by == "day" else 7  # skip by 1 or 7 days
+        usage_agg = [(None, 0)] * 24  # (sum, count) for each hour
+        min_date = timezone.localtime().date()
+        max_date = timezone.localtime().date() - datetime.timedelta(days=unit * (per_hour - 1))
+
+        for i in range(per_hour):
+            curr = date - datetime.timedelta(days=i * unit)
+            usage = self.get_usage_on_date(room, curr, field)  # usage for curr
+            # incoporate usage safely considering None (no data) values
+            usage_agg = [
+                (self.safe_add(acc[0], val), acc[1] + (1 if val is not None else 0))
+                for acc, val in zip(usage_agg, usage)
+            ]
+            # update min and max date if any data was logged
+            if any(usage):
+                min_date = min(min_date, curr)
+                max_date = max(max_date, curr)
+
+        ret = [x[0] / x[1] if x[0] and x[1] else 0 for x in usage_agg]
+        return ret, min_date, max_date
+
+    def get(self, request, room_id):
+        """
+        GET: returns the usage in terms of count or capacity of a fitness center for a given date
+        per hour aggregated by day or week for a given number of days
+        """
+
+        room = get_object_or_404(FitnessRoom, pk=room_id)
+        try:
+            if (date := request.query_params.get("date")) is None:
+                date = timezone.localtime().date()
+            else:
+                date = timezone.make_aware(datetime.datetime.strptime(date, "%Y-%m-%d")).date()
+        except ValueError:
+            return Response({"detail": "date must be in the format YYYY-MM-DD"}, status=400)
+
+        try:
+            per_hour = int(request.query_params.get("per_hour", 1))
```

Comment:
how many data points per hour to use for aggregation. per_hour=4 and group_by=day means for each hour look at the usage in the same time in the past 4 days. per_hour=2 and group_by=week means for each hour look at the usage on the same hour and same day a week ago and 2 weeks ago.

Reply:
also if i put it with the date check, we can't have specific error message

Reply:
Ah yes, specific error message seems good since this is a pretty complex query.

And got it that makes sense. Perhaps also putting a comment to label that as well? Or maybe a rename such as `num_samples` or something similar?